src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.cc

```diff
@@ -127,29 +127,107 @@ BtreeLBAManager::get_mappings(
-	  if (pos.is_end() || pos.get_key() >= (offset + length)) {
-	    TRACET("{}~{} done with {} results",
-	           c.trans, offset, length, ret.size());
-	    return typename LBABtree::iterate_repeat_ret_inner(
+    [c, offset, length, FNAME, this](auto &btree, auto &ret) {
+      return seastar::do_with(
+	std::list<BtreeLBAMappingRef>(),
+	[offset, length, c, FNAME, this, &ret, &btree](auto &pin_list) {
+	return LBABtree::iterate_repeat(
+	  c,
+	  btree.upper_bound_right(c, offset),
+	  [&pin_list, offset, length, c, FNAME](auto &pos) {
+	    if (pos.is_end() || pos.get_key() >= (offset + length)) {
+	      TRACET("{}~{} done with {} results",
+		     c.trans, offset, length, pin_list.size());
+	      return LBABtree::iterate_repeat_ret_inner(
+		interruptible::ready_future_marker{},
+		seastar::stop_iteration::yes);
+	    }
+	    TRACET("{}~{} got {}, {}, repeat ...",
+		   c.trans, offset, length, pos.get_key(), pos.get_val());
+	    ceph_assert((pos.get_key() + pos.get_val().len) > offset);
+	    pin_list.push_back(pos.get_pin(c));
+	    return LBABtree::iterate_repeat_ret_inner(
 	      interruptible::ready_future_marker{},
-	      seastar::stop_iteration::yes);
-	  }
-	  TRACET("{}~{} got {}, {}, repeat ...",
-	         c.trans, offset, length, pos.get_key(), pos.get_val());
-	  ceph_assert((pos.get_key() + pos.get_val().len) > offset);
-	  ret.push_back(pos.get_pin(c));
-	  return typename LBABtree::iterate_repeat_ret_inner(
-	    interruptible::ready_future_marker{},
-	    seastar::stop_iteration::no);
+	      seastar::stop_iteration::no);
+	  }).si_then([this, &ret, c, &pin_list] {
+	    return _get_original_mappings(c, pin_list
+	    ).si_then([&ret](auto _ret) {
+	      ret = std::move(_ret);
+	    });
+	  });
 	});
     });
 }
 
+BtreeLBAManager::_get_original_mappings_ret
+BtreeLBAManager::_get_original_mappings(
+  op_context_t<laddr_t> c,
+  std::list<BtreeLBAMappingRef> &pin_list)
+{
+  return seastar::do_with(
+    lba_pin_list_t(),
+    [this, c, &pin_list](auto &ret) {
+    return trans_intr::do_for_each(
+      pin_list,
+      [this, c, &ret](auto &pin) {
+	LOG_PREFIX(BtreeLBAManager::get_mappings);
+	if (pin->get_raw_val().is_laddr()) {
+	  TRACET(
+	    "getting original mapping for indirect mapping {}~{}",
+	    c.trans, pin->get_key(), pin->get_length());
+	  return this->get_mappings(
+	    c.trans, pin->get_raw_val().get_laddr(), pin->get_length()
+	  ).si_then([&pin, &ret, c](auto new_pin_list) {
+            LOG_PREFIX(BtreeLBAManager::get_mappings);
+	    assert(new_pin_list.size() == 1);
+	    auto &new_pin = new_pin_list.front();
+	    auto intermediate_key = pin->get_raw_val().get_laddr();
+	    assert(!new_pin->is_indirect());
+	    assert(new_pin->get_key() <= intermediate_key);
+	    assert(new_pin->get_key() + new_pin->get_length() >=
+	    intermediate_key + pin->get_length());
+            if (!ret.empty()) {
+              if (new_pin->get_key() == ret.back()->get_key()) {
```

Comment:
For the indirection mappings:
* [LA ~ 4KB] => [LIA ~ 4KB]
* [LA+4KB ~ 4KB] => [LIA+4KB ~ 4KB]
* [LIA-4KB ~ 16KB] => [PADDR ~ 16KB]

When query the range [LA ~ 8KB], does it return only one indirect-pin with key=LA, length=4KB, indirect-key=LIA, indirect-base=LIA-4KB, value=PADDR ?

Reply:
> For the indirection mappings:
> 
> * [LA ~ 4KB] => [LIA ~ 4KB]
> * [LA+4KB ~ 4KB] => [LIA+4KB ~ 4KB]
> * [LIA-4KB ~ 16KB] => [PADDR ~ 16KB]
> 
> When query the range [LA ~ 8KB], does it return only one indirect-pin with key=LA, length=4KB, indirect-key=LIA, indirect-base=LIA-4KB, value=PADDR ?

No, it will return two indirect pins that have the same indirect-key. This if condition is always false, I'll remove it.